File: app/src/main/java/ml/bmlzootown/hydravion/RequestTask.java
package ml.bmlzootown.hydravion;

import android.content.Context;

import com.android.volley.AuthFailureError;
import com.android.volley.Request;
import com.android.volley.RequestQueue;
import com.android.volley.Response;
import com.android.volley.VolleyError;
import com.android.volley.toolbox.StringRequest;
import com.android.volley.toolbox.Volley;

import java.util.HashMap;
import java.util.Map;

public class RequestTask {

    private static String response;
    private Context context;

    public RequestTask(Context context) {
        this.context = context;
    }

    public void sendRequest(String uri, final String cookies, final VolleyCallback callback) {
        RequestQueue queue = Volley.newRequestQueue(this.context);
        StringRequest stringRequest = new StringRequest(Request.Method.GET, uri,
                response -> {
                    //Log.d("JSON", response);
                    callback.onSuccess(response);
                }, new Response.ErrorListener() {
            @Override
            public void onErrorResponse(VolleyError error) {
                error.printStackTrace();
                callback.onError(error);
            }
        }) {
            @Override
            public Map<String, String> getHeaders() throws AuthFailureError {
                Map<String, String> params = new HashMap<>();
                params.put("Cookie", cookies);
                params.put("Accept", "application/json");
                return params;
            }
        };

        queue.add(stringRequest);
    }

    public void sendRequest(String uri, final String cookies, final String creatorGUID, final VolleyCallback callback) {
        RequestQueue queue = Volley.newRequestQueue(this.context);
        StringRequest stringRequest = new StringRequest(Request.Method.GET, uri,
                new Response.Listener<String>() {
                    @Override
                    public void onResponse(String response) {
                        //Log.d("JSON", response);
                        callback.onSuccessCreator(response, creatorGUID);
                    }
                }, new Response.ErrorListener() {
            @Override
            public void onErrorResponse(VolleyError error) {
                error.printStackTrace();
                callback.onError(error);
            }
        }) {
            @Override
            public Map<String, String> getHeaders() throws AuthFailureError {
                Map<String, String> params = new HashMap<>();
                params.put("Cookie", cookies);
                params.put("Accept", "application/json");
                return params;
            }
        };

        queue.add(stringRequest);
    }

    public interface VolleyCallback {

        void onSuccess(String string);

        void onSuccessCreator(String string, String creatorGUID);

        void onError(VolleyError error);
    }
}

File: app/src/main/java/ml/bmlzootown/hydravion/client/RequestTask.kt
package ml.bmlzootown.hydravion.client

import android.content.Context
import com.android.volley.*
import com.android.volley.toolbox.JsonRequest
import com.android.volley.toolbox.StringRequest
import com.android.volley.toolbox.Volley

class RequestTask(context: Context) {

    private val volleyQueue: RequestQueue = Volley.newRequestQueue(context)

    fun getReponseStatus(uri: String?, callback: VolleyCallback) {
        var responseCode: Int = 0;
        val stringRequest: StringRequest = object : StringRequest(Method.GET, uri,
            Response.Listener { _: String? ->
                //Log.d("JSON", response);
                callback.onResponseCode(responseCode);
            }, Response.ErrorListener { error: VolleyError ->
                //error.printStackTrace()
                callback.onResponseCode(error.networkResponse.statusCode)
            }) {
            override fun parseNetworkResponse(response: NetworkResponse?): Response<String>? {
                responseCode = response?.statusCode ?: 0;
                return super.parseNetworkResponse(response);
            }
        }
        stringRequest.setShouldCache(false)
        volleyQueue.add(stringRequest)
    }

    fun sendRequest(uri: String?, cookies: String, callback: VolleyCallback) {
        val stringRequest: StringRequest = object : StringRequest(Method.GET, uri,
            Response.Listener { response: String? ->
                //Log.d("JSON", response);
                callback.onSuccess(response ?: "")
            }, Response.ErrorListener { error: VolleyError ->
                error.printStackTrace()
                callback.onError(error)
            }) {
            override fun getHeaders(): Map<String, String> {
                val params: MutableMap<String, String> = HashMap()
                params["Cookie"] = cookies
                params["Accept"] = "application/json"
                params["User-Agent"] = "Hydravion (AndroidTV), CFNetwork"
                return params
            }
        }
        volleyQueue.add(stringRequest)
    }

    fun sendData(uri: String?, cookies: String, params: Map<String, String>?, callback: VolleyCallback) {
        val stringRequest: StringRequest = object : StringRequest(
            Method.POST,
            uri,
            Response.Listener { response: String? ->
                callback.onSuccess(response ?: "")
            }, Response.ErrorListener { error ->
                error.printStackTrace()
                callback.onError(error)
            }) {

            @Throws(AuthFailureError::class)
            override fun getParams(): Map<String, String>? = params

            @Throws(AuthFailureError::class)
            override fun getHeaders(): Map<String, String> =
                mapOf(
                    "Cookie" to cookies,
                    "Accept" to ACCEPT_JSON,
                    "User-Agent" to "Hydravion (AndroidTV), CFNetwork"
                )
        }
        volleyQueue.add(stringRequest)
    }

    fun sendDataWithBody(uri: String?, cookies: String, body: String, callback: VolleyCallback) {
        val jsonRequest: JsonRequest<String> = object : JsonRequest<String>(
            Method.POST,
            uri,
            body,
            Response.Listener { response: String? ->
                callback.onSuccess(response ?: "")
            }, Response.ErrorListener { error ->
                error.printStackTrace()
                callback.onError(error)
            }) {

            @Throws(AuthFailureError::class)
            override fun getHeaders(): Map<String, String> =
                mapOf(
                    "Cookie" to cookies,
                    "Accept" to ACCEPT_JSON,
                    "User-Agent" to "Hydravion (AndroidTV), CFNetwork",
                    "Content-Type" to "application/json"
                )

            override fun getBodyContentType(): String = "application/json"

            override fun parseNetworkResponse(response: NetworkResponse?): Response<String> =
                Response.success(String(response?.data ?: ByteArray(0)), null)
        }
        volleyQueue.add(jsonRequest)
    }

    fun sendRequest(uri: String?, cookies: String, creatorGUID: String?, callback: VolleyCallback) {
        val stringRequest: StringRequest = object : StringRequest(
            Method.GET,
            uri,
            Response.Listener { response: String? ->
                callback.onSuccessCreator(response ?: "", creatorGUID ?: "")
            }, Response.ErrorListener { error: VolleyError ->
                error.printStackTrace()
                callback.onError(error)
            }) {

            override fun getHeaders(): Map<String, String> =
                mapOf(
                    "Cookie" to cookies,
                    "Accept" to ACCEPT_JSON,
                    "User-Agent" to "Hydravion (AndroidTV), CFNetwork"
                )
        }
        volleyQueue.add(stringRequest)
    }

    interface VolleyCallback {
        fun onResponseCode(response: Int)
        fun onSuccess(response: String)
        fun onSuccessCreator(response: String, creatorGUID: String)
        fun onError(error: VolleyError)
    }

    companion object {

        private const val ACCEPT_JSON = "application/json"
    }
}